File: bongf/java/src/level2/OpenChatRoom.java
package level2;

import java.util.*;

public class OpenChatRoom {
    public static void main(String[] args) {
        OpenChatRoom openChatRoom = new OpenChatRoom();
        System.out.println(Arrays.toString(openChatRoom.solution(new String[]{"Enter uid1234 Muzi", "Enter uid4567 Prodo","Leave uid1234","Enter uid1234 Prodo","Change uid4567 Ryan"})));
    }

    public String[] solution(String[] record) {
        Map<String, String> dic = new HashMap<>();

        for(String r : record) {
            String[] s = r.split(" ");
            if(!s[0].equals("Leave")) {
                dic.put(s[1], s[2]);
            }
        }

        List<String> answer = new ArrayList<>();

        String[] printer = new String[]{"님이 들어왔습니다.", "님이 나갔습니다."};

        for(String r : record) {
            String[] s = r.split(" ");
            if(s[0].equals("Enter")) {
                answer.add(dic.get(s[1]) + printer[0]);
            } else if (s[0].equals("Leave")) {
                answer.add(dic.get(s[1]) + printer[1]);
            }
        }
        return answer.toArray(new String[0]);
    }
}

File: programmers/python/level2/open_chat_room.py
def solution(record):
    nickname = {}
    data = []
    
    for r in record :
        r = r.split(" ")
        if(r[0] == "Enter"):
            data.append([r[1], "님이", " 들어왔습니다."])
            nickname[r[1]] = r[2]
        elif(r[0] == "Leave"):
            data.append([r[1], "님이", " 나갔습니다."])
        else :
            nickname[r[1]] = r[2]
    
    answer = []
    for d in data :
        d[0] = nickname[d[0]]
        answer.append("".join(d))
    
    return answer

record = ["Enter uid1234 Muzi", "Enter uid4567 Prodo","Leave uid1234","Enter uid1234 Prodo","Change uid4567 Ryan"]
print(solution(record))

## printer를 써거 깔끔하게 
def solution(record):
    answer = []
    namespace = {}
    printer = {'Enter':'님이 들어왔습니다.', 'Leave':'님이 나갔습니다.'}
    for r in record:
        rr = r.split(' ')
        if rr[0] in ['Enter', 'Change']:
            namespace[rr[1]] = rr[2]

    for r in record:
        if r.split(' ')[0] != 'Change':
            answer.append(namespace[r.split(' ')[1]] + printer[r.split(' ')[0]])

    return answer

